code/vocabulary.py

```diff
@@ -90,3 +97,86 @@
 # for quick verification of the reload
 # print(str(ctx.settings["dictate.word_map"]))
 # print(str(ctx.lists["user.vocabulary"]))
+
+
+class PhraseReplacer:
+    """Utility for replacing phrases by other phrases inside text or word lists.
+
+    Replacing longer phrases has priority.
+
+    Args:
+      - phrase_dict: dictionary mapping recognized/spoken forms to written forms
+    """
+
+    def __init__(self, phrase_dict: Dict[str, str]):
+        # Index phrases by first word, then number of subsequent words n_next
+        phrase_index = dict()
+        for spoken_form, written_form in phrase_dict.items():
+            words = spoken_form.split()
+            first_word, n_next = words[0], len(words) - 1
+            phrase_index.setdefault(first_word, dict())
+            same_first_word = phrase_index[first_word]
+            same_first_word.setdefault(n_next, dict())
+            same_first_word[n_next][tuple(words[1:])] = written_form
+
+        # Sort n_next index so longer phrases have priority
+        self.phrase_index = {
+            first_word: dict(sorted(same_first_word.items(), key=lambda x: -x[0]))
+            for first_word, same_first_word in phrase_index.items()
+        }
+
+    def replace_phrases(self, input_text: Union[str, list, tuple]):
+        """Return input_text with phrases replaced"""
+        got_string = isinstance(input_text, str)
+        if got_string:
+            input_text = input_text.split()
+        input_words = tuple(input_text)
+
+        output_words = []
+        first_word_i = 0
+        while first_word_i <= len(input_words) - 1:
+            first_word = input_words[first_word_i]
+            next_word_i = first_word_i + 1
+            # Could this word be the first of a phrase we should replace?
+            for n_next, phrases_n_next in self.phrase_index.get(first_word, dict()).items():
```

Comment:
In order for this to prefer replacing longer phrases, you seem to be relying on the iteration order being the same as the order you used when you made the dictionary on line 123-126. I don't think this is guaranteed (https://docs.python.org/2/library/stdtypes.html#dict.items), especially not across different Python implementations (this matters - Talon used to use PyPy!).

It's probably better to just use `sorted()` here to sort the relevant potential matches by length. Or you could use a list of pairs instead of a dictionary for the second-level index.

Reply:
Changed to using an OrderedDict now -- should be as explicit as sorting during iteration, but faster for large vocabularies.

You're right that ordered dictionary iteration used to be a cPython implementation detail, but it was [blessed](https://docs.python.org/3/whatsnew/3.7.html#summary-release-highlights) into the language spec in python 3.7. OrderedDict still has advantages though (e.g. https://stackoverflow.com/questions/50872498).